lib/src/dense_set.rs

```diff
@@ -0,0 +1,1282 @@
+        let offset = get_offset(reg_index);
+
+        self.data[data_index] &= !(1 << offset);
+    }
+
+    fn is_empty(&self) -> bool {
+        for i in 0..self.data.len() {
+            if self.data[i] != 0 {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn intersect(&mut self, other: &Self) {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+        let mut res = Vec::<Reg>::new();
+
+        for &item in self.reg_vec.iter() {
+            if other.contains(item) {
+                res.push(item);
+            }
+        }
+        self.reg_vec = res;
+
+        for i in 0..smallest_set_size {
+            self.data[i] &= other.data[i];
+        }
+
+        for i in smallest_set_size..self.data.len() {
+            self.data[i] = 0;
+        }
+    }
+
+    fn union(&mut self, other: &Self) {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+        let greatest_set_size = cmp::max(self.data.len(), other.data.len());
+
+        for &item in other.reg_vec.iter() {
+            if !self.contains(item) {
+                self.reg_vec.push(item);
+            }
+        }
+
+        for i in 0..smallest_set_size {
+            self.data[i] |= other.data[i];
+        }
+
+        if other.data.len() > self.data.len() {
+            for i in smallest_set_size..greatest_set_size {
+                self.data.push(other.data[i]);
+            }
+        }
+    }
+
+    fn remove(&mut self, other: &Self) {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+
+        for &item in other.reg_vec.iter() {
+            if self.contains(item) {
+                self.vec_delete(item);
+            }
+        }
+
+        for i in 0..smallest_set_size {
+            self.data[i] &= !other.data[i];
+        }
+    }
+
+    fn intersects(&self, other: &Self) -> bool {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+
+        for i in 0..smallest_set_size {
+            if self.data[i] & other.data[i] != 0 {
+                return true;
+            }
+        }
+        false
+    }
+
+    fn is_subset_of(&self, other: &Self) -> bool {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+
+        for i in 0..smallest_set_size {
+            if (self.data[i] | other.data[i]) != other.data[i] {
+                return false;
+            }
+        }
+
+        if self.data.len() > other.data.len() {
+            for i in other.data.len()..self.data.len() {
+                if self.data[i] != 0 {
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    fn equals(&self, other: &Self) -> bool {
+        self.data == other.data
+    }
+
+    fn vec_get_item(&self, index: usize, real: bool) -> Option<&Reg> {
+        for reg in self.reg_vec.iter() {
+            if reg.is_real() == real && reg.get_index() == index {
```

Comment:
(And we could avoid a linear search here too.)

Reply:
> (And we could avoid a linear search here too.)

Currently the `reg_vec` is unsorted so you mean by using the `Vec<Option<Reg>>` implementation (with having the same index for the `Reg` and the `Vec`) ?

Reply:
The linear search is avoided in ff39c7c93a66c9ecced9bf607a1935b38ab31c51 because of Vec<Option<Reg>> but the overall time for register allocation is multiplied by 3.